criu/tls.c

```diff
@@ -0,0 +1,336 @@
+	}
+
+	return ret;
+}
+
+/*
+ * Read data from a file descriptor, then encrypt and send it with GnuTLS.
+ * This function is used for cases when we would otherwise use splice()
+ * to transfer data from PIPE to TCP socket.
+ */
+ssize_t tls_send_data_from_fd(int fd, unsigned long len)
+{
+	ssize_t copied;
+	ssize_t total_sent = 0;
+	char *buffer = xmalloc(len);
+
+	if (!buffer)
+		return -1;
+
+	while (len > total_sent) {
+		copied = read(fd, (buffer + total_sent), (len - total_sent));
+		if (copied < 0) {
+			pr_perror("Can't read from pipe");
+			total_sent = -1;
+			goto err;
+		}
+
+		if (copied == 0)
+			break;
+
+		int sent = 0;
+		while (sent < copied) {
+			int ret = tls_send((buffer + total_sent + sent), (copied - sent));
+
+			if (ret < 0) {
+				tls_perror("Failed sending data", sent);
+				total_sent = -1;
+				goto err;
+			}
+
+			sent += ret;
+		}
+
+		total_sent += copied;
+	}
+err:
+	xfree(buffer);
+	return total_sent;
+}
+
+ssize_t tls_recv(void *buf, size_t len)
+{
+	int ret = gnutls_record_recv(session, buf, len);
+
+	if (ret == GNUTLS_E_REHANDSHAKE) {
+		tls_perror("Re-handshake is not supported", ret);
+		return -1;
+	}
+
+	if (gnutls_error_is_fatal(ret)) {
+		tls_perror("Failed to read data", ret);
+		return -1;
+	}
+
+	return ret;
+}
+
+/*
+ * Read and decrypt data with GnuTLS and write it to a file descriptor.
+ * This function is used for cases when we would otherwise use splice()
+ * to transfer data from a TCP socket to a PIPE.
+ */
+ssize_t tls_recv_data_to_fd(int fd, unsigned long len)
+{
+	ssize_t received;
+	ssize_t total_written = 0;
+	char *buffer = xmalloc(len);
+
+	if (!buffer)
+		return -1;
+
+	while (len > total_written) {
+		received = tls_recv((buffer + total_written), (len - total_written));
+		if (received < 0) {
+			total_written = -1;
+			goto err;
+		}
+
+		int written = 0;
+		while (written < received) {
+			int ret = write(fd,
+				(buffer + total_written + written),
+				(received - written));
+
+			if (ret < 0) {
+				pr_perror("Failed writing to fd");
+				total_written = -1;
+				goto err;
+			}
+
+			written += ret;
+		}
+
+		total_written += received;
+	}
+err:
+	xfree(buffer);
+	return total_written;
+}
+
+static int tls_set_priority(gnutls_session_t *session)
+{
+	int ret;
+
+	ret = gnutls_priority_init(&priority_cache, TLS_PRIORITY, NULL);
+	if (ret == GNUTLS_E_SUCCESS)
+		ret = gnutls_priority_set(*session, priority_cache);
+
+	return ret;
+}
+
+static ssize_t _tls_data_push_cb(gnutls_transport_ptr_t ptr, const void* data, size_t len)
+{
+	int sockfd = *(int *)(ptr);
+	return send(sockfd, data, len, 0);
+}
+
+static ssize_t _tls_data_pull_cb(gnutls_transport_ptr_t ptr, void* data, size_t maxlen)
+{
+	int sockfd = *(int *)(ptr);
+	return recv(sockfd, data, maxlen, 0);
+}
+
+static int tls_handshake()
+{
+	int ret;
+
+	ret = tls_set_priority(&session);
+	if (ret != GNUTLS_E_SUCCESS) {
+		tls_perror("Failed to set priority", ret);
+		return -1;
+	}
+
+	connfdPtr = xmalloc(sizeof(int));
```

Comment:
xmalloc() can fail and leak on the error-path.
But, do you need to allocate `connfdPtr`?
Won't `gnutls_transport_set_ptr(session, (gnutls_transport_ptr_t)&tls_sk)` work?